src/testnet/node.rs

```diff
@@ -0,0 +1,385 @@
+        let chain_tip = match self.bootstraping_chain {
+            true => StacksHeaderInfo::genesis(),
+            false => match &self.chain_tip {
+                Some(chain_tip) => chain_tip.clone(),
+                None => unreachable!()
+            }
+        };
+
+        // Constructs the coinbase transaction - 1st txn that should be handled and included in 
+        // the upcoming tenure.
+        let coinbase_tx = self.generate_coinbase_tx();
+
+        // Construct the upcoming tenure
+        let tenure = LeaderTenure::new(
+            chain_tip, 
+            self.average_block_time,
+            coinbase_tx,
+            self.config.clone(),
+            self.mem_pool.clone(),
+            microblock_secret_key, 
+            sortitioned_block.clone(),
+            vrf_proof);
+
+        Some(tenure)
+    }
+
+    /// Handles artefacts coming from an ongoing tenure.
+    /// At this point, we're not updating the chainstate, we're simply having the node
+    /// candidating for the next tenure.
+    pub fn receive_tenure_artefacts(&mut self, anchored_block_from_ongoing_tenure: &StacksBlock, parent_block: &SortitionedBlock) {
+        let ops_tx = self.burnchain_ops_tx.take().unwrap();
+
+        if self.active_registered_key.is_some() {
+            let registered_key = self.active_registered_key.clone().unwrap();
+
+            let vrf_proof = self.keychain.generate_proof(
+                &registered_key.vrf_public_key, 
+                parent_block.sortition_hash.as_bytes()).unwrap();
+
+            let op = self.generate_block_commit_op(
+                anchored_block_from_ongoing_tenure.header.block_hash(),
+                1, // todo(ludo): fix
+                &registered_key, 
+                &parent_block,
+                VRFSeed::from_proof(&vrf_proof));
+
+            ops_tx.send(op).unwrap();
+        }
+
+        // Naive implementation: we keep registering new keys
+        let vrf_pk = self.keychain.rotate_vrf_keypair();
+        let op = self.generate_leader_key_register_op(vrf_pk, parent_block.consensus_hash); // todo(ludo): should we use the consensus hash from the burnchain tip, or last sortitioned block?
+        ops_tx.send(op).unwrap();
+
+        self.burnchain_ops_tx = Some(ops_tx);
+    }
+
+    /// Process artefacts from the tenure.
+    /// At this point, we're modifying the chainstate, and merging the artifacts from the previous tenure.
+    pub fn process_tenure(&mut self, anchored_block: &StacksBlock, parent_block: &SortitionedBlock, microblocks: Vec<StacksMicroblock>, burn_db: Arc<Mutex<BurnDB>>) {
+
+        {
+            let mut db = burn_db.lock().unwrap();
+
+            let mut tx = db.tx_begin().unwrap();
+
+            let res = self.chain_state.preprocess_anchored_block(
+                &mut tx, 
+                &parent_block.burn_header_hash,
+                &anchored_block, 
+                &parent_block.parent_burn_header_hash).unwrap();
+        }
```

Comment:
You're also going to want to call `preprocess_streamed_microblock()` for each microblock.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1202/commits/f47f032c213247937e50c82002e605531b66659a